File: app/src/main/java/de/datlag/hotdrop/view/animation/CircularAnimation.java
package de.datlag.hotdrop.view.animation;

import android.view.View;
import android.view.animation.Animation;
import android.view.animation.Transformation;

public class CircularAnimation extends Animation {

    private View view;
    private float cx, cy;
    private float prevX, prevY;
    private float r;
    private float prevDx, prevDy;

    public CircularAnimation(View view, float r){
        this.view = view;
        this.r = r;
    }

    public boolean willChangeBounds() {
        return false;
    }

    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        int cxImage = width / 2;
        int cyImage = height / 2;
        cx = view.getLeft() + cxImage;
        cy = view.getTop() + cyImage;

        prevX = cx;
        prevY = cy;
    }

    protected void applyTransformation(float interpolatedTime, Transformation t) {
        if(interpolatedTime == 0){
            t.getMatrix().setTranslate(prevDx, prevDy);
            return;
        }

        float angleDeg = (interpolatedTime * 360f + 90) % 360;
        float angleRad = (float) Math.toRadians(angleDeg);

        // r = radius, cx and cy = center point, a = angle (radians)
        float x = (float) (cx + r * Math.cos(angleRad));
        float y = (float) (cy + r * Math.sin(angleRad));

        float dx = prevX - x;
        float dy = prevY - y;

        prevX = x;
        prevY = y;

        prevDx = dx;
        prevDy = dy;

        t.getMatrix().setTranslate(dx, dy);
    }
}

File: app/src/main/java/de/datlag/hotdrop/view/animation/CircularAnimation.kt
package de.datlag.hotdrop.view.animation

import android.view.View
import android.view.animation.Animation
import android.view.animation.Transformation
import kotlin.math.cos
import kotlin.math.sin

class CircularAnimation(private val view: View, private val r: Float) : Animation() {
    private var cx = 0f
    private var cy = 0f
    private var prevX = 0f
    private var prevY = 0f
    private var prevDx = 0f
    private var prevDy = 0f
    override fun willChangeBounds(): Boolean {
        return false
    }

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        val cxImage = width / 2
        val cyImage = height / 2
        cx = view.left + cxImage.toFloat()
        cy = view.top + cyImage.toFloat()
        prevX = cx
        prevY = cy
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
        if (interpolatedTime == 0f) {
            t.matrix.setTranslate(prevDx, prevDy)
            return
        }
        val angleDeg = (interpolatedTime * 360f + 90) % 360
        val angleRad = Math.toRadians(angleDeg.toDouble()).toFloat()
        val x = (cx + r * cos(angleRad.toDouble())).toFloat()
        val y = (cy + r * sin(angleRad.toDouble())).toFloat()
        val dx = prevX - x
        val dy = prevY - y
        prevX = x
        prevY = y
        prevDx = dx
        prevDy = dy
        t.matrix.setTranslate(dx, dy)
    }

}